src/v/cluster/tm_stm.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/persisted_stm.h"
+#include "config/configuration.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "raft/consensus.h"
+#include "raft/errc.h"
+#include "raft/logger.h"
+#include "raft/state_machine.h"
+#include "raft/types.h"
+#include "storage/snapshot.h"
+#include "utils/expiring_promise.h"
+#include "utils/mutex.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include <compare>
+
+namespace cluster {
+
+struct tm_etag {
+    int64_t log_etag{0};
+    int64_t mem_etag{0};
+
+    tm_etag inc_mem() const {
+        return tm_etag{
+          .log_etag = this->log_etag, .mem_etag = this->mem_etag + 1};
+    }
+
+    tm_etag inc_log() const {
+        return tm_etag{.log_etag = this->log_etag + 1, .mem_etag = 0};
```

Comment:
is the intention here to return `.mem_etag = 0` or `this->mem_etag` (which would be symmetric with `inc_mem` above).

Reply:
right, incrementing log version (major component) and resetting minor